src/components/Dropdown/Dropdown.tsx

```diff
@@ -0,0 +1,209 @@
+  const [lastSelected, setLastSelected] = useState(-1);
+  const [listOffset, setListOffset] = useState(0);
+  const dropdownRef = useRef<HTMLDivElement | null>(null);
+  const dropdItemRef = useRef<HTMLButtonElement | null>(null);
+  const inputRef = useRef<HTMLInputElement | null>(null);
+  const itemHeight = 36;
+  const dropdownHeight = 300;
+  const noOfItems = dropdownHeight / itemHeight;
+
+  const assertVisible = (index: number) => {
+    if (dropdownRef.current && dropdItemRef.current) {
+      const offset = (index - (noOfItems - 2)) * itemHeight;
+
+      if (index === suggestions.length - 1) {
+        setListOffset(offset > 0 ? -offset + itemHeight : 0);
+      } else {
+        setListOffset(offset > 0 ? -offset : 0);
+      }
+    }
+  };
+
+  useEffect(() => {
+    function handleClickOutside(event: any) {
+      if (dropdownRef.current
+        && !dropdownRef.current.contains(event.target as Node)
+      ) {
+        setPending(true);
+      }
+    }
+
+    document.addEventListener('mousedown', handleClickOutside);
+
+    return () => {
+      document.removeEventListener('mousedown', handleClickOutside);
+    };
+  }, [dropdownRef]);
+
+  const setSuggestionsDebounced = useCallback(
+    debounce(() => {
+      if (inputRef?.current?.value) {
+        const input = inputRef.current.value.toLowerCase();
+
+        setSuggestions(
+          words.filter(word => word.toLowerCase().includes(input)),
+        );
+      } else {
+        setSuggestions([]);
+      }
+
+      assertVisible(-1);
+      setSelectedSug(-1);
+      setPending(false);
+    }, delay),
+    [],
+  );
+
+  const handleInputChange = useCallback(
+    () => {
+      setPending(true);
+      setSuggestionsDebounced();
+    }, [],
+  );
+
+  const handleKeyDown = useCallback(
+    (e: React.KeyboardEvent<HTMLInputElement>) => {
+      switch (e.key) {
+        case 'ArrowDown':
+          if (selectedSug < suggestions.length - 1) {
+            assertVisible(selectedSug + 1);
+            setSelectedSug(prev => prev + 1);
+          } else {
+            assertVisible(0);
+            setSelectedSug(0);
+          }
+
+          break;
+        case 'ArrowUp':
+          if (selectedSug > 0) {
+            assertVisible(selectedSug - 1);
+            setSelectedSug(prev => prev - 1);
+          } else {
+            assertVisible(suggestions.length - 1);
+            setSelectedSug(suggestions.length - 1);
+          }
+
+          break;
+        case 'Enter':
+          if (inputRef?.current?.value) {
+            if (selectedSug === -1) {
+              if (suggestions.includes(inputRef.current.value)) {
+                save(inputRef.current.value);
+                inputRef.current.value = '';
+              }
+            } else {
+              inputRef.current.value = suggestions[selectedSug];
+              save(suggestions[selectedSug]);
+              assertVisible(-1);
+              setSelectedSug(-1);
+            }
+          }
```

Comment:
needs some refactoring. Seems that you can create a couple of variables and make it inside one `if`